libraries/extractor/src/main/java/androidx/media3/extractor/Ac4Util.java

```diff
@@ -113,13 +210,335 @@ private SyncFrameInfo(
+              int nSkipBytes = dataBitArray.readBits(7);  // n_skip_bytes
+              for (int j = 0; j < nSkipBytes; j++) {
+                dataBitArray.skipBits(8);
+              }
+              break;
+          }
+        }
+        ac4Presentation.preVirtualized = dataBitArray.readBit();  // b_pre_virtualized
+        bAddEmdfSubstreams = dataBitArray.readBit();  // b_add_emdf_substreams
+      }
+      if (bAddEmdfSubstreams) {
+        int nAddEmdfSubstreams = dataBitArray.readBits(7);  // n_add_emdf_substreams
+        for (int j = 0; j < nAddEmdfSubstreams; j++) {
+          dataBitArray.skipBits(5 + 10);  // substream_emdf_version and substream_key_id
+        }
+      }
+
+      boolean bPresentationBitrateInfo = false;
+      if (presentationVersion > 0) {
+        bPresentationBitrateInfo = dataBitArray.readBit();  // b_presentation_bitrate_info
+      }
+
+      if (bPresentationBitrateInfo) {
+        if (!parseDsiBitrate(dataBitArray)) {
+          throw ParserException.createForUnsupportedContainerFeature(
+              "Can't parse bitrate DSI.");
+        }
+      }
+
+      if (presentationVersion > 0) {
+        boolean bAlternative = dataBitArray.readBit();  // b_alternative
+        if (bAlternative) {
+          dataBitArray.byteAlign();
+          int nameLen = dataBitArray.readBits(16);  // name_len
+          byte[] presentationName = new byte[nameLen];
+          dataBitArray.readBytes(presentationName, 0, nameLen);
+          ac4Presentation.description = ByteBuffer.wrap(presentationName);
+
+          int nTargets = dataBitArray.readBits(5);  // n_targets
+          for (int i = 0; i < nTargets; i++) {
+            dataBitArray.skipBits(3);  // target_md_compat
+            dataBitArray.skipBits(8);  // target_device_category
+          }
+        }
+      }
+
+      dataBitArray.byteAlign();
+
+      if (ac4DsiVersion == 1) {
+        long end = (dsiSize - dataBitArray.bitsLeft()) / 8;
+        long presentationBytes = end - start;
+        if (presBytes < presentationBytes) {
+          throw ParserException.createForUnsupportedContainerFeature(
+              "pres_bytes is smaller than presentation_bytes.");
+        }
+        long skipBytes = presBytes - presentationBytes;
+        dataBitArray.skipBits((int)skipBytes * 8);
+      }
+
+      // We should know this or something is probably wrong
+      // with the bitstream (or we don't support it)
+      if (ac4Presentation.channelCoded && ac4Presentation.channelMode == -1) {
+        throw ParserException.createForUnsupportedContainerFeature(
+            "Can't determine channel mode of presentation " + presentationIndex);
+      }
+
+      ac4Presentations.put(presentationIndex, ac4Presentation);
+    }
+
+    int channelCount = -1;
+    // Using first presentation (default presentation) channel count
+    int presentationIndex = 0;
+    Ac4Presentation ac4Presentation =
+        Objects.requireNonNull(ac4Presentations.get(presentationIndex));
```

Comment:
@ybai001,

Why are we looping over all `presentationIndex` above and storing it in a `Map` when all we require is the `ac4Presentation` for the index `0`?

Is it to verify if the file doesn't have bad data? Even in that case storing it in a `Map` seems an overkill, when we don't intend to use the map for anything else but index `0`.

Reply:
@rohitjoins, 
This AC4 parser implementation is similar as the one in Android Framework: https://cs.android.com/android/platform/superproject/main/+/main:frameworks/av/media/module/extractors/mp4/AC4Parser.cpp;drc=2e87c835abea90b0cf61b42dc9aa8a882f2ef5aa;l=292. It parses all the information for all presentations because there is API https://developer.android.com/reference/android/media/MediaExtractor#getAudioPresentations(int) to fetch the presentations info for appointed audio track.

Since AndroidX Media extractor doesn't support AudioPresentation, I know this loop is redundant now. 

But I know Android framework MPEGExtractor was deprecated and it will be replaced by AndroidX Media extractor. Although AndroidX Media doesn't support AudioPresentation now but I believe it will support it sooner or later. That's why I use a Map to store all audio presentation info and keep this logic in my implementation.

If one audio track includes multiple audio presentations, and if application doesn't set which presentation should be played, the Dolby AC-4 decoder will decode the default presentation (the first presentation).

Reply:
@ybai001,

I would suggest to remove any logic which will be used in the future as it can be added when required. For now we should only parse and use the `first presentation`.

Reply:
@rohitjoins , OK, I'll update the code.

Reply:
@rohitjoins, do you have any update on this pull request?

Reply:
Hi @ybai001,

I'll take a look and send this PR for internal review.

Reply:
Hi @rohitjoins ,

Do you have further updates on this CL. We have more support added for AC-4 presentations parsing for different containers that are dependent on this pull request. I am looking to create new pull requests once this pull request is reviewed and merged.

Thanks!

Reply:
Hi, @rohitjoins,

Any update on this PR for internal review? Thanks.

Reply:
Hi @rohitjoins,

Just wondering if this CL is good to go. Please let us know if you need any further clarification needed regarding the changes here.

We have few other dependent changes in the pipeline that needs to be pushed up for review once this CL gets merged.
 
Thanks

Reply:
> Hi @sqshet-dlb,
> 
> I’m currently making some updates to the PR, such as adding more documentation and refactoring some parts into new methods before sending it for internal review. I’ll keep you updated on the progress and aim to merge it soon.

Hi, @rohitjoins, as you mentioned on July 2nd,
_I'll take a look and send this PR for internal review._
It's nearly 3 months. Could you help to explain why this pull request has not been sent for internal review? Could you give me a deadline? This is very important for us since it blocks several other contributions as @sqshet-dlb said.

Reply:
Hi, @rohitjoins 
Could you help to provide us any update about this contribution? 
Thanks.

Reply:
@rohitjoins , there is a merge conflict with IAMF code. Do I need to resolve this conflict? Because you said you would send this for internal code review, I'm not sure whether I should modify code again. Thanks.

Reply:
Hey @ybai001,

No need to update this PR as I already do it for the internal version. If no blockers come up in review, this should be merged this week.

Reply:
@rohitjoins, thanks for the quick feedback. :)